lib/middleware.js

```diff
@@ -36,29 +35,14 @@ function commandMiddlewareFactory (commandMiddleware) {
   })
 }
 
-function applyMiddleware (argv, yargs, middlewares, beforeValidation) {
-  const beforeValidationError = new Error('middleware cannot return a promise when applyBeforeValidation is true')
-  return middlewares
-    .reduce((accumulation, middleware) => {
-      if (middleware.applyBeforeValidation !== beforeValidation) {
-        return accumulation
-      }
-
-      if (isPromise(accumulation)) {
-        return accumulation
-          .then(initialObj =>
-            Promise.all([initialObj, middleware(initialObj, yargs)])
-          )
-          .then(([initialObj, middlewareObj]) =>
-            Object.assign(initialObj, middlewareObj)
-          )
-      } else {
-        const result = middleware(argv, yargs)
-        if (beforeValidation && isPromise(result)) throw beforeValidationError
-
-        return isPromise(result)
-          ? result.then(middlewareObj => Object.assign(accumulation, middlewareObj))
-          : Object.assign(accumulation, result)
-      }
-    }, argv)
+async function applyMiddleware (argv, yargs, middlewares, beforeValidation) {
+  for (let middleware of middlewares.filter(m => m.applyBeforeValidation === beforeValidation)) {
+    Object.assign(argv,
+      middleware.length < 3
+        ? await middleware(argv, yargs)
+        // middleware using a done callback
+        : await new Promise(resolve => middleware(argv, yargs, resolve))
```

Comment:
here the same. You've missed error rejection

Reply:
You're right. The same logic is to be applied to builders, handlers, middlewares and completion functions.

Reply:
done